### PR TITLE
AOTAutograd: fix 'Trying to backward through the graph a second time' error

### PR DESCRIPTION
**Pull request by bdhirsh**
Fixes https://github.com/pytorch/pytorch/issues/97745. See discussion and comment in the PR for more details.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98848
* #98031
* #98030
* __->__ #98960

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by bertmaher — APPROVED** [review]

File: test/dynamo/test_repros.py [hunk 1]
```diff
@@ -902,6 +902,31 @@ def f(x):
...
+        a_ref = torch.randn(2, 2, requires_grad=True)
+        b_ref = torch.randn(2, 2, requires_grad=True)
+        out_ref = f(a_ref, b_ref)
+
+        a_test = a_ref.clone().detach().requires_grad_(True)
+        b_test = b_ref.clone().detach().requires_grad_(True)
+        out_test = torch.compile(f, backend="aot_eager")(a_test, b_test)
+
+        self.assertEqual(out_ref, out_test)
+        print(a_ref.grad)
```
Comment (gchanan) [inline test/dynamo/test_repros.py:925]: probably don't want these prints in the test?

File: torch/_functorch/aot_autograd.py [hunk 2]
```diff
@@ -2322,6 +2329,66 @@ def aot_dispatch_autograd(flat_fn, flat_args: List[Any], aot_config: AOTConfig,
...
+        #     propagating gradients
+        # By the time we're executing a node and we see that it produces a None, the set of nodes to execute
+        # is already locked-in.
+        #
+        # The fix: instead, we can recognize statically that the graph we're compiling will never contribute
+        # gradients to y, and prevent autograd from trying to traverse y's autograd tape at all.
+        # We can do this by manually detach'ing y before sending it through the `CompiledFunction`.
+        #
+        # Note that this solution is not bulletproof.
+        # It's possible to construct a case where eager may or may have have tried to autograd through y,
```
Comment (albanD) [inline torch/_functorch/aot_autograd.py:2371]: missing "not" ?

**Review by albanD — APPROVED** [review]
Very nice comment!

**Comment by bdhirsh** [comment]
Looked at the failures, it looks like this PR regresses this case:
```
import torch
import torch.nn as nn


@torch.compile(backend='aot_eager')
def f(x):
    return torch.masked.norm(x, 0, mask=None)

a = torch.tensor(0.8604, requires_grad=True)
out = f(a)
out.sum().backward()
```

Digging into it more. also fixing the lint

**Comment by bdhirsh** [comment]
So, it looks like there are some cases where we legitimately assign a None gradient for a few ops.. e.g. `norm()` ops, here: https://github.com/pytorch/pytorch/blob/c650d7b67fdc7594c7c400721f481914aac51e6e/torch/csrc/autograd/FunctionsManual.cpp#L265

So in eager mode, the above repro prints the following:
```
import torch
import torch.nn as nn


def f(x):
    return torch.masked.norm(x, 0, mask=None)

a = torch.tensor(0.8604, requires_grad=True)
out = f(a)
out.sum().backward()
print(a.grad)  # prints None!
```

cc @albanD does this sound like a bug or a feature (and would it diverge here and raise an error with torch.compile)? I'm not sure that we can differentiate between this case and the detach() case, since in both cases our compiled bw graph will produce `None` grad_inputs.

**Comment by bdhirsh** [comment]
some offline discussion: @gchanan pointed out that in theory we could update these derivative formulas to return zero tensors instead of None, and use EfficientZeroTensor to avoid the perf penalty. @albanD pointed out that if the zero tensor needs to get populated in the grad field, then we're forced to materialize the tensor of zeros (expensive)

For now in this PR, I'm xfail'ing the tests above and living with the (slight) difference in eager mode. To be clear, this error only shows up if we have a graph such that **every** tensor input that requires gradients would have gotten a None grad in eager mode.

File: test/dynamo/test_dynamic_shapes.py [hunk 3]
```diff
@@ -31,6 +31,8 @@
     "ReproTests": [
         # Could not infer dtype of torch._C.SymIntNode
         "test_convert_boxes_to_pooler_format",
+        # TODO: look into this: FakeTensor error.
+        "test_gan_repro_trying_to_backward_through_the_graph_a_second_time",
```
Comment (bdhirsh) [inline test/dynamo/test_dynamic_shapes.py:35]: The error I hit was very weird, and seemed unrelated to the actual issue in this PR. Ignoring so we can land this PR asap, and I'll come back to it later.

**Comment by bdhirsh** [comment]
`linalg.det` also has the same issue (was failing an inductor test) - I skipped the test in the same way. We might want to consider fixing `linalg.det` to not return `None` gradients though (in this case it's cheap, since it only appears to happen when you input is size `(0, 0)`

```
>>> a = torch.randn(0, 0, requires_grad=True, device='cuda')
>>> torch.linalg.det(a).sum().backward()
>>> a.grad
>>>
```
